### PR TITLE
Extend dataset search and add search API endpoint

### PR DESCRIPTION
**Pull request by mvdbeek**
The goal is to be able to search for datasets and dataset collections via the API, without knowing which history a dataset is in. This also extends the `HistoryContentsFilters` in a way that allows searching for datasets or collections with a given tag or annotation.

### CODE REVIEW

**Comment by martenson** [comment]
Would this replace the current approach in view_multiple histories page?

**Comment by mvdbeek** [comment]
I think it might, but I haven't yet looked at what the client-side code currently does. This should definitely be re-usable, and filtering by tag or datatype for instance should be easy to add to the current standard history view.

**Comment by mvdbeek** [comment]
I think this works OK now, but it looks like the most interesting searches (in the history and multi-history views) are client-only searches ? The saved dataset search is still using controllers, so I'm not sure if this is worth investing energy in. Are we OK with having this exist as a API-only function at this point ? 
I'd really like to have this because I occasionally need to find some old dataset buried in a collection, and being able to search by tool id would save me a lot of time.

**Comment by mvdbeek** [comment]
@galaxybot test py3

File: test/api/test_datasets.py [hunk 1]
```diff
@@ -3,21 +3,93 @@
...
 
 from base import api
-from base.populators import TestsDatasets
+from base.populators import (
+    DatasetCollectionPopulator,
+    DatasetPopulator,
+)
 
 
-class DatasetsApiTestCase(api.ApiTestCase, TestsDatasets):
```
Comment (jmchilton) [inline test/api/test_datasets.py:9]: Awesome, been meaning to do that forever.

**Comment by jmchilton** [comment]
It is absolutely fine IMO to have API-only functionality. The client-side multi-history search I think is a dead-end and I wouldn't worry too much about it.

**Comment by martenson** [comment]
@jmchilton I agree with both of those statements. @mvdbeek thanks!

**Pull request merged**